### PR TITLE
Merge fixes for JENKINS-31258 and JENKINS-31524

### PR DESCRIPTION
**Pull request by demonfiddler**
Please accept these fixes:
- JENKINS-31258 (Jenkins Maven plug-in ignores JUnit-format test results from unknown Maven plug-ins)
- JENKINS-31524 (SurefireArchiver ignores updated results when multiple testing plug-ins use the same reports directory).

Thanks!

### CODE REVIEW

**Comment by demonfiddler** [comment]
Sorry, looks like my PR message got muddled somehow. The PR is to merge my fixes for JENKINS-31258 and JENKINS-31524 *to* jenkinsci (not *from*).

File: src/main/resources/hudson/maven/Messages.properties [hunk 1]
```diff
@@ -55,7 +55,7 @@ MavenProbeAction.DisplayName=Monitor Maven Process
 MavenProcessFactory.ClassWorldsNotFound=No classworlds*.jar found in {0} -- Is this a valid maven directory?
 
 MavenRedeployer.DisplayName=Deploy to Maven repository
-MavenVersionCallable.MavenHomeDoesntExist=Maven Home {0} doesn\u2019t exist
+MavenVersionCallable.MavenHomeDoesntExist=Maven Home {0} doesn't exist
```
Comment (daniel-beck) [inline src/main/resources/hudson/maven/Messages.properties]: This used to be http://www.fileformat.info/info/unicode/char/2019/index.htm which is better than what you propose, why the change?

Comment (daniel-beck) [inline src/main/resources/hudson/maven/Messages.properties]: Nevermind, "*Fixed a compilation failure that was apparently caused an extraneous (and inconsistent!) Unicode right-apostrophe character…*"

Comment (demonfiddler) [inline src/main/resources/hudson/maven/Messages.properties]: @daniel-beck I found that this would not load under Windows 8, which I traced to the use of the deprecated Unicode right-apostrophe character \u2019. Then I discovered that the ordinary ASCII apostrophe had been used elsewhere in this and other properties files, so it seemed logical to enforce consistency and fix the Windows build in one change! Hope this explains and justifies the change :-)

File: src/main/java/hudson/maven/reporters/SurefireArchiver.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ public boolean postExecute(MavenBuildProxy build, MavenProject pom, MojoInfo moj
                 fileSet = Iterables.filter(fileSet, new Predicate<File>() {
                     @Override
                     public boolean apply(File input) {
-                        return !parsedFiles.containsKey(input);
+                        return !parsedFiles.containsKey(input) || parsedFiles.get(input) < input.lastModified();
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/SurefireArchiver.java]: I doubt it works reliably in all cases, because the file can be technically replaced by an older one (e.g. move operation). I would expect something like Map<File, Set<Long>>, where the set stores modification times or even MD5s of the files.

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 3]
```diff
@@ -167,21 +168,34 @@ public boolean canRunTests(MojoInfo mojo) {
...
+     * Returns the location of test reports created by the specified MOJO.
+     * @param pom The project model.
+     * @param mojo The MOJO.
+     * @return The directory containing the test reports.
+     * @throws ComponentConfigurationException if unable to retrieve the report directory from the MOJO configuration.
+     */
+    protected File getReportsDirectory(MavenProject pom, MojoInfo mojo) throws ComponentConfigurationException {
+    	// [JENKINS-31258] Allow unknown MOJOs to contribute test results in arbitrary locations by setting a Maven property.
+		String reportsDirectoryOverride = getReportsDirectoryOverride(mojo);
+		if (reportsDirectoryOverride != null)
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Tabs instead of spaces

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 4]
```diff
@@ -167,21 +168,34 @@ public boolean canRunTests(MojoInfo mojo) {
...
+			return mojo.expressionEvaluator.alignToBaseDirectory(new File(reportsDirectoryOverride));
+
         if (this.reportDirectoryConfigKey != null) {
             File reportsDir = mojo.getConfigurationValue(this.reportDirectoryConfigKey, File.class);
-            if (reportsDir != null && reportsDir.exists()) {
-                return getReportFiles(reportsDir, getFileSet(reportsDir));
-            } 
-            
+            if (reportsDir != null)
+                return reportsDir;
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: This plugin is a splitt-off of the core, hence it follows the Oracle codestyle. In this style brackets are required for single-liners

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 5]
```diff
@@ -167,21 +168,34 @@ public boolean canRunTests(MojoInfo mojo) {
...
+    }
+
+    /**
+     * Returns the location of test reports created by the specified MOJO.
+     * @param pom The project model.
+     * @param mojo The MOJO.
+     * @return The directory containing the test reports.
+     * @throws ComponentConfigurationException if unable to retrieve the report directory from the MOJO configuration.
+     */
+    protected File getReportsDirectory(MavenProject pom, MojoInfo mojo) throws ComponentConfigurationException {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: From what I see this method never returns null
Maybe makes sense to mark it as <code>@Nonnull</code> and to remove the check above

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 6]
```diff
@@ -231,12 +245,41 @@ public static TestMojo lookup(String artifactId, String groupId, String goal) {
     
     public static TestMojo lookup(MojoInfo mojo) {
         TestMojo testMojo = lookup(mojo.pluginName.groupId, mojo.pluginName.artifactId, mojo.getGoal());
+
+        if (testMojo == null && getReportsDirectoryOverride(mojo) != null) {
+			testMojo = FALLBACK;
+        }
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: tabs

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 7]
```diff
@@ -231,12 +245,41 @@ public static TestMojo lookup(String artifactId, String groupId, String goal) {
...
+     * @param mojo An unknown MOJO.
+     * @return the value of the expression <code>${jenkins.&lt;mojo-execution-id&gt;.reportsDirectory}</code>.
+     */
+    private static String getReportsDirectoryOverride(MojoInfo mojo) {
+        // Validity of the override property should be constrained to "test" and "integration-test" phases but there seems to be no
+        // way to check this, as MojoExecution.getLifecyclePhase() (added 2009-05-12) causes tests to throw NoSuchMethodError!!!
+        // So, we're obliged to assume that the property is configured for a valid test life cycle phase.
+//        String phase = mojo.mojoExecution.getLifecyclePhase();
+//        if ("test".equals(phase) || "integration-test".equals(phase)) {
+        try {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Commented code. Requires TODO or cleanup

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 8]
```diff
@@ -231,12 +245,41 @@ public static TestMojo lookup(String artifactId, String groupId, String goal) {
...
         return null;
     }
-    
+
+    /**
+     * For an unknown test-capable MOJO, returns the path to its reports directory as configured by a Maven property.
+     * @param mojo An unknown MOJO.
+     * @return the value of the expression <code>${jenkins.&lt;mojo-execution-id&gt;.reportsDirectory}</code>.
+     */
+    private static String getReportsDirectoryOverride(MojoInfo mojo) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: <code>@CheckForNull</code> + Javadoc for this case?

File: src/main/resources/hudson/maven/Messages.properties (hunk 1 above)
Comment (daniel-beck) [inline src/main/resources/hudson/maven/Messages.properties]: @demonfiddler  Sure, thanks. I just hadn't seen the commit comment I quoted yet. What I wonder is how and by whom was that character deprecated?

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 9]
```diff
@@ -231,12 +245,41 @@ public static TestMojo lookup(String artifactId, String groupId, String goal) {
...
+        // So, we're obliged to assume that the property is configured for a valid test life cycle phase.
+//        String phase = mojo.mojoExecution.getLifecyclePhase();
+//        if ("test".equals(phase) || "integration-test".equals(phase)) {
+        try {
+            String reportsDirExpr = "${jenkins." + mojo.mojoExecution.getExecutionId() + ".reportsDirectory}";
+            Object result = mojo.expressionEvaluator.evaluate(reportsDirExpr);
+            if (result != null && !result.equals(reportsDirExpr) && !result.toString().trim().isEmpty())
+                return result.toString().trim();
+        } catch (ExpressionEvaluationException e) {
+            // Ignore evaluation exceptions.
```
Comment (oleg-nenashev) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Maybe makes sense to log the warning

File: src/test/java/hudson/maven/MojoInfoBuilder.java [hunk 10]
```diff
@@ -44,12 +57,22 @@ public MojoInfoBuilder version(String version) {
...
+        return this;
+    }
+    
     public MojoInfoBuilder startTime(long startTime) {
         this.startTime = startTime;
         return this;
     }
     
-    public MojoInfoBuilder configValue(String key,String value) {
+    public MojoInfoBuilder configValue(String key, String value) {
```
Comment (oleg-nenashev) [inline src/test/java/hudson/maven/MojoInfoBuilder.java]: please avoid unrelated changes. they may impact merges due to parallel notifications

**Comment by oleg-nenashev** [comment]
Do you plan to develop a test for JENKINS-31524? I'm not sure it's mandatory, but it may be useful of course

File: src/main/java/hudson/maven/reporters/SurefireArchiver.java (hunk 2 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/SurefireArchiver.java]: The pasedFiles Map<File, Long> (no longer a Set) does indeed store the modification timestamp of any pre-existing file - that was the essence of this change. Since the files in question are test results I think it's rather unlikely that they would have anything than a current timestamp.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 3 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: This file has used spaces throughout since its inception by Christoph Kutzinski on 06/01/2013. The Jenkins 'code of conduct' requests that one mainain stylistic consistency and explicitly bans auto-reformatting.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 4 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Fair enough, but how does one know what the local coding standards are? I didn't see a .checkstyle or IDE project file anywhere. I will reinstate the brackets.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 5 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Yes, I think you're right. Will do as suggested.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 6 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Ditto

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 7 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: But what would the TODO action be? The code comment is attempting to explain what it would prefer to do at this point and the reason why it cannot do so. It's unclear to me why the commented code compiles but won't execute. If I remove this block of code altogether we've lost track of the important point that we ought really to constrain executions to testing phases.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 8 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Will do.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 9 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: Will do.

File: src/test/java/hudson/maven/MojoInfoBuilder.java (hunk 10 above)
Comment (demonfiddler) [inline src/test/java/hudson/maven/MojoInfoBuilder.java]: Sorry, incorrect spacing/punctuation is one of my pet hates ;-)

**Comment by demonfiddler** [comment]
Re. a test case for JENKINS-31524 I certainly can do, if you think it's appropriate/desirable.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 8 above)
Comment (imod) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: is this anywhere documented for the enduser? e.g. in a help

File: src/main/resources/hudson/maven/Messages.properties (hunk 1 above)
Comment (demonfiddler) [inline src/main/resources/hudson/maven/Messages.properties]: I was just going on the build error message (which I'm afraid I don't recall) followed by various web search results. The resulting web pages variously seemed be saying that \u2019 is an incorrect/obsolete/ambiguous code point for the right apostrophe / right single quote character and there was some question as to the languages in which its use is appropriate. Anyway, it was the only way I could find to make the build run on Windows, plus this properties file now consistently uses standard ASCII single quote characters (\u0027) throughout.

File: src/main/java/hudson/maven/reporters/TestMojo.java (hunk 8 above)
Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: I'm not sure where one would put such documentation. https://wiki.jenkins-ci.org/display/JENKINS/Building+a+maven2+project perhaps?

**Comment by demonfiddler** [comment]
I don't understand what's failed here. The CloudBees job (#180) shows 205 (+2) tests, 2 skipped (as before), 0 failures.

**Comment by daniel-beck** [comment]
@jenkinsci/code-reviewers The author requested a code review a while back on the dev list.

File: src/main/java/hudson/maven/reporters/TestMojo.java [hunk 11]
```diff
@@ -231,12 +247,44 @@ public static TestMojo lookup(String artifactId, String groupId, String goal) {
...
         }
+
         return null;
     }
-    
+
+    /**
+     * For an unknown test-capable MOJO, returns the path to its reports directory as configured by a Maven property.
+     * @param mojo An unknown MOJO.
+     * @return the value of the expression <code>${jenkins.&lt;mojo-execution-id&gt;.reportsDirectory}</code>.
```
Comment (imod) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: is this expression documented for the enduser? I was not able to find it anywhere...

Comment (demonfiddler) [inline src/main/java/hudson/maven/reporters/TestMojo.java]: @imod, I was planning to document the new Jenkins-Maven property format in the [Building a maven2 project Wiki page](https://wiki.jenkins-ci.org/display/JENKINS/Building+a+maven2+project), but naturally this can only be done after my Pull Request has been merged and the new plug-in version formally adopted by Jenkins Releng. Please advise if there is some other mechanism for documenting such changes?

**Comment by oleg-nenashev** [comment]
restarting the build

**Comment by demonfiddler** [comment]
15 weeks elapsed since PR submitted, 4 weeks since successful test build... any chance of this getting accepted?

**Comment by demonfiddler** [comment]
Thanks for the merge @olamy, Wiki updated with a description of the new mechanism.
